File: 77.组合.cpp
/*
 * @lc app=leetcode.cn id=77 lang=cpp
 *
 * [77] 组合
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        if (k <= 0) {
            return res;
        }
        vector<int> candidates;
        for (int i = 0; i < n; ++i) {
            candidates.push_back(i+1);
        }
        vector<int> current;
        foo(candidates, 0, k, current, res);
        return res;
    }
    void foo(const vector<int>& candidates, 
            int start_idx, int total, const vector<int>& current, 
            vector<vector<int>>& res) {
        if (total == 0) { 
            res.push_back(current);
            return;
        }
        if (start_idx >= candidates.size()) {
            return;
        }
        for (int i = start_idx; i < candidates.size(); ++i) {
            auto next = current;
            next.push_back(candidates[i]);
            foo(candidates, i + 1, total - 1, next, res);
        }
        return;
    }
};
// @lc code=end

File: 77.组合.py
#
# @lc app=leetcode.cn id=77 lang=python
#
# [77] 组合
#

# @lc code=start
class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        def foo(target, num, cur, res):
            if num == 0:
                res.append(cur)
                return

            for i in target:
                tmp = target[:]
                tmp.remove(i)
                if not cur or i > cur[-1]:
                    foo(tmp, num - 1, cur + [i], res)

        res = []
        foo(range(1, n+1), k, [], res)
        return res
# @lc code=end

